### PR TITLE
Added rviz plugins

### PR DESCRIPTION
**Pull request by NovoG93**
To keep the vision_msgs package agnostic I added a meta package and moved initial vision_msgs package

### CODE REVIEW

File: vision_msgs_meta/CMakeLists.txt
Comment (SteveMacenski) [inline vision_msgs_meta/CMakeLists.txt]: No meta package is required, lets keep them separated. You can make the rviz plugins though tied to the msgs (as would be needed) so they'd come in pairs if you grabbed the rviz plugins.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
You have the plugins as a submodule, they should be given as raw source to be adopted here

File: Readme.md
Comment (SteveMacenski) [inline Readme.md]: Move the readme from the vision_msgs package into the root after this reorganization and add a section about the rviz plugins + that lovely gif you made

Comment (SteveMacenski) [reply Readme.md]: Remove this file's contents

File: vision_msgs/vision_msgs/create_aabb.hpp
Comment (SteveMacenski) [inline vision_msgs/vision_msgs/create_aabb.hpp]: Is the static really necessary? That makes this have threading issues if called in multiple threads at high frequency in a processing pipeline

Comment (SteveMacenski) [reply vision_msgs/vision_msgs/create_aabb.hpp]: ditto below. Inline should suffice

Comment (NovoG93) [reply vision_msgs/vision_msgs/create_aabb.hpp]: I did not touch those files. They had the static already.

File: Readme.md
Comment (NovoG93) [reply Readme.md]: Implemented the structure changes

File: vision_msgs/LICENSE
Comment (SteveMacenski) [inline vision_msgs/LICENSE]: I believe the root license is sufficient when matched with the package.xml declarations of Apache 2.0

File: vision_msgs_rviz_plugins/LICENSE
Comment (SteveMacenski) [inline vision_msgs_rviz_plugins/LICENSE]: ditto

File: vision_msgs_rviz_plugins/README.md
Comment (SteveMacenski) [inline vision_msgs_rviz_plugins/README.md]: Remove this, it should be in binary format and this isn't any different than any other ROS 2 package

File: vision_msgs_rviz_plugins/README.md
Comment (SteveMacenski) [inline vision_msgs_rviz_plugins/README.md]: Some? This looks like all :wink:

File: vision_msgs_rviz_plugins/package.xml
Comment (SteveMacenski) [inline vision_msgs_rviz_plugins/package.xml:7]: Go ahead and put me too so I can get build farm failure messages if there are problems

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Otherwise, seems good from a first pass. One thing to do is make sure you lint the code via `ament_uncrustify` (hint `--reformat` will do alot of the work for you, but you may need to check on its line extensions, it can break in odd places sometimes) and `ament_cpplint` for ROS styling. I see you have some files of 4 spaces and others of 2 spaces for indentation (2 is ROS style). Consistency is key to not exploding my brain while trying to quickly scan code during maintaining processes :smile:  

After linting, I can look at the code parts in more detail, but I don't see anything obvious or significant wrong

File: vision_msgs_rviz_plugins/README.md
Comment (NovoG93) [reply vision_msgs_rviz_plugins/README.md]: Was thinking of adding BoundingBox2D/Array in the future when I need it for my project

File: vision_msgs_rviz_plugins/package.xml
Comment (SteveMacenski) [inline vision_msgs_rviz_plugins/package.xml]: I don't think this is a rosdep key

File: vision_msgs_rviz_plugins/package.xml
Comment (SteveMacenski) [inline vision_msgs_rviz_plugins/package.xml]: I don't think this is a rosdep key, I think there's a yaml vendor

File: vision_msgs_rviz_plugins/CMakeLists.txt
Comment (SteveMacenski) [inline vision_msgs_rviz_plugins/CMakeLists.txt]: again, I think this is a vendor

File: vision_msgs_rviz_plugins/package.xml
Comment (SteveMacenski) [reply vision_msgs_rviz_plugins/package.xml]: These are important because when I try to release this as a binary, the build farm will fail to find the dependency and fail the job. The package.xml having all the dependencies + be properly named is critical

File: vision_msgs_rviz_plugins/src/bounding_box_3d.cpp
Comment (SteveMacenski) [inline vision_msgs_rviz_plugins/src/bounding_box_3d.cpp]: Are you the open source robotics foundation and/or is this code heavily based on code that is copyrighted to them?

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Sorry for the multiple rounds, I'm really not trying to be pedantic here

File: vision_msgs_rviz_plugins/src/bounding_box_3d.cpp
Comment (NovoG93) [reply vision_msgs_rviz_plugins/src/bounding_box_3d.cpp]: No I am not. I thought since I was contributing here it had to be that way?
It's the first time that I contribute to an open source project so I am not sure about the procedure.

Should I put my name?

File: vision_msgs_rviz_plugins/package.xml
Comment (NovoG93) [reply vision_msgs_rviz_plugins/package.xml]: Don't worry about the multiple rounds. I am sorry that it is needed.

The rosdep keys should be valid since:

1. I build everything in a docker (ros:humble-ros-core-jammy) and ran `rosdep install --from-paths src --ignore-src -r -y` without error.
2. During earlier commits the [build on the ros buildfarm failed](https://build.ros2.org/job/Hpr__vision_msgs__ubuntu_jammy_amd64/6/console#console-section-9) if it could not reslove the rosdep keys. Here it could not find _qt5-default_. On the latest commit the buildfarm did not throw any errors and could resolve the keys:
Screenshot from 2023-01-27 20-17-04

File: vision_msgs_rviz_plugins/CMakeLists.txt
Comment (NovoG93) [reply vision_msgs_rviz_plugins/CMakeLists.txt]: Should I choose this implementation instead: https://github.com/ros2/yaml_cpp_vendor ?

File: vision_msgs_rviz_plugins/src/bounding_box_3d.cpp
Comment (SteveMacenski) [reply vision_msgs_rviz_plugins/src/bounding_box_3d.cpp]: No, this should be you if you're contributing on your own time and resources - or your company otherwise

In general, unless you sign away your rights (commonly known as a Contributor License Agreement, or CLA), when you contribute to an open source project, you still own the copyright but you're giving royalty free and irrevocable control over the code to the repository under the terms of the software license. A CLA might exist so that a repository can re-license the code or do other legal things to it to make it easier. Else, to relicense, you'd need every single copyright holder's approval to relicense (which is unlikely to occur due to sheer logistics). So in many ways, not having a CLA is a nice protection in open-source against someone doing something against the spirit of open source since I, as the maintainer, can't simply wave a wand and start charging people to use this code that you wrote

Comment (NovoG93) [reply vision_msgs_rviz_plugins/src/bounding_box_3d.cpp]: Thank you for pointing that out to me.
Will change it with the next push

File: vision_msgs_rviz_plugins/package.xml
Comment (SteveMacenski) [reply vision_msgs_rviz_plugins/package.xml]: https://github.com/ros/rosdistro/blob/c95e19fa3c38ca124982f66db59f907fecc3d6aa/foxy/distribution.yaml#L2505 is preferred so that it can build on other non-linux OS's as well since ROS 2 is supposed to support WIndows and MacOS. I'm aware of this from experience, the yaml-cpp key is only linux

File: vision_msgs_rviz_plugins/CMakeLists.txt
Comment (SteveMacenski) [reply vision_msgs_rviz_plugins/CMakeLists.txt]: Yeah! See comment above

File: vision_msgs_rviz_plugins/package.xml
Comment (SteveMacenski) [reply vision_msgs_rviz_plugins/package.xml]: tf_transformations is definitely not right though

Comment (SteveMacenski) [reply vision_msgs_rviz_plugins/package.xml]: That's not the "real" TF, that's David Lu's custom library which I honestly don't know much about

**Review by SteveMacenski — COMMENTED** [review]
Once CI turns over + that new TF key, we should be good to go

File: vision_msgs_rviz_plugins/package.xml
Comment (NovoG93) [reply vision_msgs_rviz_plugins/package.xml]: The tf_transformations is basically a wrapper around the new transforms3d that provides seamless  backward compability. As far as I understand the tf_transformations is also a valid rosdep key since it is in this repo: https://github.com/ros/rosdistro/blob/master/humble/distribution.yaml#L6301
and can be installed via apt as: ros-humble-tf-transformations

Comment (NovoG93) [reply vision_msgs_rviz_plugins/package.xml]: My latest push includes now the yaml_cpp_vendor: https://github.com/NovoG93/vision_msgs/blob/ros2/vision_msgs_rviz_plugins/package.xml#L15
Should be fine now?

Comment (SteveMacenski) [reply vision_msgs_rviz_plugins/package.xml]: I guess I've very much prefer not introducing something like that into an interface visualization library since we don't know how long if / when David might stop releasing it to new distributions. TF we know will be around for as long as ROS is around and can have the same type of function to convert RPY to quaternions https://answers.ros.org/question/69754/quaternion-transformations-in-python/ (or even just that's a trivial function to implement either way and I'm sure there are batteries included python libraries that deal with that kind of thing)

Comment (NovoG93) [reply vision_msgs_rviz_plugins/package.xml]: Okay, now I understand why you want me to switch.
Will do so👍

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
Thanks @NovoG93 for the work and great gifs / demos! This will be very nice to have out of the box. I'll plan on releasing this at the end of next week in my previously scheduled release cycle!

**Comment by SteveMacenski** [comment]
You may want to post on ROS Discourse about this (with your gif!) for visibility

**Comment by NovoG93** [comment]
Glad it finally worked out!
Will post on ROS Discourse once it is released.

**Comment by SteveMacenski** [comment]
@NovoG93 We have some build failures that will stop me from running binaries until we have them resolved. I was planning on syncing this on Friday so if you could patch up things today, tomorrow or Thursday we can keep on track

```
12:32:59 In file included from /tmp/ws/build_isolated/vision_msgs_rviz_plugins/include/vision_msgs_rviz_plugins/../../../../src/vision_msgs/vision_msgs_rviz_plugins/include/vision_msgs_rviz_plugins/detection_3d_array.hpp:33,
12:32:59                  from /tmp/ws/build_isolated/vision_msgs_rviz_plugins/include/vision_msgs_rviz_plugins/moc_detection_3d_array.cpp:9:
12:32:59 /tmp/ws/src/vision_msgs/vision_msgs_rviz_plugins/include/vision_msgs_rviz_plugins/detection_3d_common.hpp: In member function ‘void rviz_plugins::Detection3DCommon<MessageType>::updateColorConfig()’:
12:32:59 /tmp/ws/src/vision_msgs/vision_msgs_rviz_plugins/include/vision_msgs_rviz_plugins/detection_3d_common.hpp:488:20: error: ‘rcpputils::fs’ has not been declared
12:32:59   488 |     if (rcpputils::fs::exists(tmp_path)) {
12:32:59       |                    ^~
12:32:59 make[2]: *** [CMakeFiles/vision_msgs_rviz_plugins.dir/build.make:113: CMakeFiles/vision_msgs_rviz_plugins.dir/include/vision_msgs_rviz_plugins/moc_detection_3d_array.cpp.o] Error 1
```

https://build.ros2.org/job/Fdev__vision_msgs__ubuntu_focal_amd64/18/console

At the very least, that package isn't linked against or included in the package.xml

**Comment by NovoG93** [comment]
@SteveMacenski will look into that now

**Comment by NovoG93** [comment]
@SteveMacenski  I just noticed that the jenkins build job is running on Ubuntu20.04 with Foxy.
Will the vision_msgs be migrated from humble to foxy as they are different (e.g. Detection3D).
The rviz plugin in the PR is only compatible with the vision_msgs definition of Humble, for Foxy I only implemented a small part of the Humble capabilities (https://github.com/NovoG93/vision_msgs_rviz_plugins/tree/foxy) however this code was not included in the PR.

Besides the _rcpputils::fs_ problem there was a problem with the QT Meta-Object Compiler on the common headers that is also fixed with the latest commit.  
For testing I used the docker images _ros:rolling-ros-core-jammy_, _ros:humble-ros-core-jammy_, and _ros:foxy-ros-core-focal_ and built the package using: `. /opt/ros/$ROS_DISTRO/setup.sh && PYTHONIOENCODING=utf_8 PYTHONUNBUFFERED=1 colcon build --build-base build_isolated --install-base install_isolated --test-result-base test_results --event-handlers console_cohesion+ --cmake-args -DBUILD_TESTING=0` as it is apparently executed in the [jenkins build job](https://build.ros2.org/job/Fdev__vision_msgs__ubuntu_focal_amd64/18/consoleFull#console-section-12)   



I created a new PR for that here: https://github.com/ros-perception/vision_msgs/pull/84

**Comment by SteveMacenski** [comment]
It is actually released to Humble, Rolling, etc. The foxy job is just the one that's failing currently.

**Pull request merged**
